node/src/main/scala/coop/rchain/node/runtime/NodeRuntime.scala

```diff
@@ -212,284 +249,4 @@ class NodeRuntime[F[_]: Monixable: ConcurrentEffect: Parallel: Timer: ContextShi
-      override def runCancelable[A](fa: TaskEnv[A])(
-          cb: Either[Throwable, A] => IO[Unit]
-      ): SyncIO[CancelToken[TaskEnv]] =
-        t.runCancelable(envToTask(fa))(cb).map(taskToEnv(_))
-      override def runAsync[A](
-          fa: TaskEnv[A]
-      )(cb: Either[Throwable, A] => IO[Unit]): SyncIO[Unit] =
-        t.runAsync(envToTask(fa))(cb)
-      // Async
-      override def async[A](k: (Either[Throwable, A] => Unit) => Unit): TaskEnv[A] = c.async(k)
-      override def asyncF[A](k: (Either[Throwable, A] => Unit) => TaskEnv[Unit]): TaskEnv[A] =
-        c.asyncF(k)
-      // Concurrent
-      override def start[A](fa: TaskEnv[A]): TaskEnv[Fiber[TaskEnv, A]] = c.start(fa)
-      override def racePair[A, B](
-          fa: TaskEnv[A],
-          fb: TaskEnv[B]
-      ): TaskEnv[Either[(A, Fiber[TaskEnv, B]), (Fiber[TaskEnv, A], B)]] = c.racePair(fa, fb)
-      override def suspend[A](thunk: => TaskEnv[A]): TaskEnv[A]          = c.defer(thunk)
-      override def bracketCase[A, B](acquire: TaskEnv[A])(use: A => TaskEnv[B])(
-          release: (A, ExitCase[Throwable]) => TaskEnv[Unit]
-      ): TaskEnv[B]                                        = c.bracketCase(acquire)(use)(release)
-      override def raiseError[A](e: Throwable): TaskEnv[A] = c.raiseError(e)
-      override def handleErrorWith[A](fa: TaskEnv[A])(f: Throwable => TaskEnv[A]): TaskEnv[A] =
-        c.handleErrorWith(fa)(f)
-      override def flatMap[A, B](fa: TaskEnv[A])(f: A => TaskEnv[B]): TaskEnv[B] =
-        c.flatMap(fa)(f)
-      override def tailRecM[A, B](a: A)(f: A => TaskEnv[Either[A, B]]): TaskEnv[B] =
-        c.tailRecM(a)(f)
-      override def pure[A](x: A): TaskEnv[A] = c.pure(x)
-    }
-
-    /**
-      * ReaderT instances for NodeRuntime dependencies. Implementations for Log and EventLog are created "manually"
-      * although they can be generated with cats.tagless @autoFunctorK macros but support is missing for IntelliJ.
-      * https://github.com/typelevel/cats-tagless/issues/60 (Cheers, Marcin!!)
-      */
-    implicit val lg: Log[TaskEnv]       = Log[F].mapK(taskToEnv)
-    implicit val el: EventLog[TaskEnv]  = EventLog[F].mapK(taskToEnv)
-    implicit val tm: Timer[TaskEnv]     = Timer[F].mapK(taskToEnv)
-    implicit val mn: Monixable[TaskEnv] = Monixable[F].mapK(taskToEnv, NodeCallCtx.init)
-
-    for {
-      id <- NodeEnvironment.create[F](nodeConf)
-
-      // Create NodeRuntime instance
-      runtime = new NodeRuntime[TaskEnv](nodeConf, kamonConf, id, scheduler)
-
-      // Run reader layer with initial state
-      _ <- runtime.main.run(NodeCallCtx.init)
-    } yield ()
-  }
-
-  trait Cleanup[F[_]] {
-    def close(): F[Unit]
-  }
-
-  def cleanup[F[_]: Sync: Log](
-      casperStoreManager: KeyValueStoreManager[F]
-  ): Cleanup[F] =
-    new Cleanup[F] {
-      override def close(): F[Unit] =
-        for {
-          _ <- Log[F].info("Shutting down Casper store manager ...")
-          _ <- casperStoreManager.shutdown
```

Comment:
Maybe I miss something but I cannot find where it is performs in new code.

Reply:
Instead of relaying to close resource at the exit which may be skipped because of an error, RNodeKeyValueStoreManager is now exposed as a Resource.
https://github.com/rchain/rchain/blob/b95bcf47b593d40757c8b3489d8fda9a929d0402/shared/src/main/scala/coop/rchain/store/KeyValueStoreManagerSyntax.scala#L33-L37

And used here to construct combined Resource with components which depends on it.
https://github.com/rchain/rchain/blob/b95bcf47b593d40757c8b3489d8fda9a929d0402/node/src/main/scala/coop/rchain/node/runtime/NodeRuntime.scala#L186-L222

The body of `Resource#use` is where all resources are available. When `use` exits all resources are ensured to be closed which means `RNodeKeyValueStoreManager#shutdown` will be called, on any error or cancellation.
https://github.com/rchain/rchain/blob/b95bcf47b593d40757c8b3489d8fda9a929d0402/node/src/main/scala/coop/rchain/node/runtime/NodeRuntime.scala#L230-L239